Implement BSIP87: Force-Settlement Fee

christophersanborn opened the pull request:
Implementation of BSIP87.

Mostly complete <s>but need container for debt asset to receive collateral-denominated fees</s>.

- [x] Collateral-asset fee container (via PR #2159)
- [x] Fee calculation and routing
- [ ] Test cases

abitmore commented:
IMO the fastest solution is to reuse the container for market fee sharing.

abitmore commented on libraries/chain/asset_evaluator.cpp:
```diff
@@ -299,6 +304,10 @@ void_result asset_update_evaluator::do_evaluate(const asset_update_operation& o)
+   FC_ASSERT( d.head_block_time() >= HARDFORK_CORE_BSIP87_TIME
+              || !o.new_options.extensions.value.force_settle_fee_percent.valid(),
+              "A BitAsset's FSFP cannot be set before Hardfork BSIP87" );
+
```
Please add checks in `proposal_create_evaluator` as well.

christophersanborn replied:
Added in cd4753b.

abitmore commented:
If we go with the approach that adding a new container into the `asset_dynamic_asset_object`, I think we can reuse the `asset_claim_fees_operation` to claim the fees in the container by adding an `asset_id` field in the `extensions`.

MichelSantos commented on libraries/chain/db_market.cpp:
```diff
@@ -911,22 +911,58 @@ bool database::fill_call_order( const call_order_object& order, const asset& pay
+   // debt asset is passed to the pay function so it knows where to put the fee. Note that
+   // amount of collateral asset upon which fee is assessed is reduced by market_fees already
+   // paid to prevent the total fee exceeding total collateral.
+   asset force_settle_fees = pay_force_settle_fees( get(pays.asset_id), receives - market_fees );
```
> IMO the force-settlement fee should be paid before the market fees, since the operation's main business is the exchanging of the debt and the collateral, and the force-settlement fee is a part of the business logic. In comparison, the market fee is much more passive here.

I agree with my reading of your comment: I do not consider force-settlement to be a market exchange because it is, in my opinion, a transaction between the settler and the blockchain.

But my opinion seems to be contradicted by the fact the _existing_ code does charge a market fee for force-settlement, which implies that the code considers the transaction to be a market exchange.  For whatever it's worth.

MichelSantos replied:
> So it really hinges on exactly what the scope of the market fee is and what activities are intended to be subjected to it.
> 
> Tagging @MichelSantos in here for some business analytic expertise. This is an interesting question.

I think that the proposed implementation is consistent with two points in the specifcation

1. The sequencing of the multiplication in the formula implies the sequencing in the proposed implementation

2. The [prose](https://github.com/bitshares/bsips/blob/master/bsip-0081.md) can be interpreted to mean that the "delta" force settlement fee is the second calculation (emphasis added)

> When a force settlement is executed, the buyer sells smartcoin with quantity `X` and get collateral in quantity `X*(1-FSO)*(1-FSFP)/feed price price*`, the settled debt position owner sells collateral in quantity `X*(1-FSO)/feed price*` and get smartcoin in quantity `X`, the **delta** and received collateral in quantity `X*FSFP*(1-FSO)/feed price*` will be paid to the owner of the smartcoin as force settlement fee.

I think that the proposed implementation is a reasonable reading of the slightly vague specification.

abitmore replied:
You convinced me. I agree that the current implementation makes sense.

abitmore commented on libraries/chain/db_market.cpp:
```diff
@@ -1293,4 +1331,31 @@ asset database::pay_market_fees(const account_object* seller, const asset_object
+   {
+      const auto& asset_dyn_data = collecting_asset.dynamic_asset_data_id(*this);
+      modify( asset_dyn_data, [&settle_fee]( asset_dynamic_data_object& obj ){
+         obj.accumulated_fees += settle_fee.amount;  // TODO: Wrong place to put fee
```
Yes, I'm aware of it.

Please keep in mind that the collateral asset type of a debt asset can be changed, so we probably need a multi-asset container for storing the fee. Another option is, only allows the collateral asset type to be changed when the container is empty.

BSIP74 needs the container as well. Please coordinate with @jmjatlanta to avoid duplicate efforts.

@dls-cipher please keep an eye on this matter.

dls-cipher replied:
@abitmore - please confirm that BSIP74 and BSIP87 have diff container types.

Thanks

abitmore replied:
@dls-cipher I think bsip74 and bsip87 can use the same container.

abitmore commented:
The code doesn't compile. It seems some new code didn't got pushed to the repository.

Another thing is we also need to charge fees for instant force-settlements (for GSed assets).

christophersanborn commented:
> The code doesn't compile. It seems some new code didn't got pushed to the repository.

Correct, it now depends on PR #2159, which is not merged in on this branch.  Question: @abitmore — If I rebase this branch on top of `hardfork` tip, (or, actually, on top of PR #2159, which is based on `hardfork` tip as of yesterday), will it erase the discussion history here?  Or would it be better to "merge" recent hard fork back into this branch?

> Another thing is we also need to charge fees for instant force-settlements (for GSed assets).

I'm not sure I agree with this.  It's not mentioned in the BSIP — and I note there is something odd about an asset owner charging fees on settlement for an asset that's in a "failed" state, as opposed to letting claimants claim their full value from the settlement pool.  It also could be a vector for an exit scam — as the asset owner could in theory ramp the fees up substantially after GS and make off with a big chunk of the pool, effectively abandoning their product and their customers.  

I would argue that the differences between "instant settling" and "force settling" are not just semantic but also substantive, and enough so to argue that instant settling is not covered by the BSIP as written.  I'm trying not to editorialize too much though. 

If I recall, but I haven't gone back to double-check, the market fees mechanism is also not triggered during instant settle, (although perhaps it should be, since they are collected by the collateral issuer, who was not responsible for the GS).  This also partly informed my choice not to touch the GS settle code.

abitmore commented:
Merging from base branch is usually better for tracking old discussions.

Fees about instant settlement was discussed here: https://github.com/bitshares/bsips/issues/260#issuecomment-611573590. You're correct that it's not explicitly described in the BSIP document. I don't think this PR is the best place to discuss it, will forward your comments to the bsips repo. IMO, when trading an asset, the traders need to trust the issuer/owner to an extent, as a platform we provide tools for both the owners and the traders E.G. asset permissions.

christophersanborn commented:
> don't think this PR is the best place to discuss it, will forward your comments to the bsips repo. 

Cool thanks for raising the point there. I'll consider it a point of active discussion and await resolution there.

abitmore commented:
> Merging from base branch is usually better for tracking old discussions.

@christophersanborn you merged from the container branch, although the container branch has been merged into the hardfork branch now, it seems Github doesn't render well but still shows the changes in the container branch, which makes it a bit unclear for code review. It's not a big deal though. If you'd like to fix it, you can recreate the branch from the commit before the merge, merge `hardfork`, then force-push (but not rebase directly. Also remember to make a backup before acting).

christophersanborn commented:
> > Merging from base branch is usually better for tracking old discussions.
> 
> @christophersanborn you merged from the container branch, although the container branch has been merged into the hardfork branch now, it seems Github doesn't render well but still shows the changes in the container branch, which makes it a bit unclear for code review. It's not a big deal though. If you'd like to fix it, you can recreate the branch from the commit before the merge, merge `hardfork`, then force-push (but not rebase directly. Also remember to make a backup before acting).

@abitmore — Cool, thanks for tip.  Looks better now.  Should be some new commits coming in over next couple of days addressing above review points and such.

abitmore commented:
Thanks. If got chance please wrap the long lines (max 118 characters per line).

christophersanborn commented on libraries/protocol/include/graphene/protocol/asset_ops.hpp:
```diff
@@ -34,6 +34,8 @@ namespace graphene { namespace protocol {
       // After BSIP81 activation, taker_fee_percent is the taker fee
       fc::optional<uint16_t>                  taker_fee_percent;
+      // MERGE NOTE: BSIP-74 additions here
+      fc::optional<uint16_t> force_settle_fee_percent;  // BSIP-87
```
Question: (@abitmore) — Have I made the right choice to add this new `force_settle_fee_percent` parameter inside the `additional_asset_options` structure?  As I’ve thought more about it, it has seemed to me it maybe should instead have been added to a new `additional_bitasset_options` structure, to become an extension to `bitasset_options` (instead of `asset_options`).  This new parameter (along with `margin_call_fee_ratio` from BSIP74) are bit-asset-specific, whereas the remaining existing parameters are generic to all coin types, I believe. 

Advantages would be: 

* Not storing parameters for asset types to which they do not apply.

* Owners would use the `asset_update_bitasset_operation` rather than `asset_update_operation` to change the parameters, which seems more contextually appropriate.

Disadvantages would be:

* Introducing a little bit more code complexity (new structures, extension types).  But, this seems reasonable to me.

* There is some precedent already — e.g. there are bitasset-specific flags and permissions which are stored in the `asset_options` even though they’d make more sense in `bitasset_options`.   

* An asset owner wishing to update all the relevant parameters would need to broadcast two separate operations, one to update the common fees and one to update the bitasset fees.

* It would require a little bit of refactoring of the code already written, but I don’t think we’ve dug ourselves too big of a hole thus far.

As mentioned, this also affects BSIP74.  When I last spoke with @jmjatlanta, (some days ago), he was supportive of moving the parameter to the new structure, assuming nothing has changed since then.

What are your thoughts? Is there a reason to keep it in the common options? Or should we move these ones over to bitasset options?

abitmore replied:
Yes, I think it's better to put those options in `bitasset_options`. There was already some work done in https://github.com/bitshares/bitshares-core/pull/2157, but I didn't merge it so far because I hoped that your pulls got merge first then I rebase. But now it seems it's better that you work on top of it, so I'll merge it, after doen please rebase or merge. Thanks.

abitmore replied:
Update: #2157 has been merged into `hardfork` branch, please merge `hardfork` into the development branch again. Thanks.

christophersanborn commented on libraries/chain/db_market.cpp:
```diff
@@ -1293,4 +1331,31 @@ asset database::pay_market_fees(const account_object* seller, const asset_object
 
+/***
+ * @brief Calculate force-settlement fee and give it to issuer of the settled asset
+ * @param collecting_asset the smart asset object which should receive the the fee
```
Fixed in cd4753b.

abitmore commented:
Merging for progress.

christophersanborn commented:
> Thanks. Code looks good. Expecting test cases.

Thanks.  I have a number of test cases already thanks to @MichelSantos.  I am just making slight adjustments to them due to moving the force_settle_fee_percent into bitasset_opts. Should be able to check them in in the morning.

christophersanborn commented:
@abitmore — I've added more commits to [`christophersanborn:cjs-bsip87`](https://github.com/christophersanborn/bitshares-core/tree/cjs-bsip87), incorporating @MichelSantos's unit tests.  But I do not see the update here, presumably because the PR has been merged.  What is the right way to proceed?  Can this PR be re-opened?  Or should I open a new one?

christophersanborn commented:
Also, one more question — a thing I want to be sure I am doing right: (@abitmore)

In `asset_update_bitasset_evaluator::do_apply()`, I’ve made no changes.  It looks to me like the existing code basically basically takes the new options and replaces the old options with them. (Along with some preconditions and postcondition trigger checks.)  And I am assuming that it is correct to leave this alone.

But If I’m interpreting the code correctly, this means that if any _optional_ parameter is unset (like the new `force_settle_fee_percent`), then it will be unset after the update is applied… even if that parameter previously had a value.

Is that the correct interpretation of unset optional fields in an update operation?  And does that match the way these are treated in other operations?

In other words, I can see two possible interpretations of an unset optional field in an update operation:

(1) I do not wish to change this parameter — after the update, it should retain the value it previously held.

(2) I wish the parameter to be unset after the update — I.e., even if a value was previously set, I want the new state to be unset.

I would be inclined to think that interpretation (1) would be the interpretation “of least surprise”, unless precedent has already established interpretation (2).

Do we need to add interpretation (1) logic to ::do_apply()?

abitmore commented:
a) For test cases, please create a new pull request.
b) For `asset_update_bitasset_operation`, IMO it's better to go interpretation (2) since it keeps the code simple, if we go (1) then we need a complex design to be able to unset the new parameters. I do think it's better to document it though. There was a similar [discussion](https://github.com/bitshares/bitshares-core/pull/838#discussion_r183513777) for the optional `target_collateral_ratio` parameter in `call_order_update_operation`.

christophersanborn commented:
> b) For `asset_update_bitasset_operation`, IMO it's better to go interpretation (2) since it keeps the code simple, if we go (1) then we need a complex design to be able to unset the new parameters when they were already set. I do think it's better to document it though. There was a similar [discussion](https://github.com/bitshares/bitshares-core/pull/838#discussion_r183513777) for the optional `target_collateral_ratio` parameter in `call_order_update_operation`.

Not sure that this generalizes to all optional parameters, but in case of `force_settle_fee_paramter`, since `unset` is interpreted as zero anyway, it is not strictly required to have an option to un-set the parameter.   

One possible middle ground would be to reject the operation in do_evaluate if an optional field is unset in the operation but set (and non-zero) in the existing object. That would force the asset owner to be explicit — and I think could have some merit.

I do agree we should not pursue interpretation (1) unless we are doing it across the board — it would be worse to have inconsistent update semantics than to have non-ideal semantics.  If we are being consistent with the rest of the API (primary point of comparison would be the existing reward_percent options in additional_asset_options), then interpretation (2) is fine by me.

abitmore commented:
Yes, I agree to keep consistent with the rest of code base. In current code base we have quite some places doing interpretation (2), so IMO it's good that we keep it as is.

abitmore commented:
@christophersanborn wrote:
> primary point of comparison would be the existing reward_percent options in additional_asset_options

According to the code, there is no special code about setting or unsetting the `reward_percent` parameter in account options.
